src/nvim/buffer.c

```diff
@@ -309,7 +309,7 @@ close_buffer (
   // buffers.
   // The caller must take care of NOT deleting/freeing when 'bufhidden' is
   // "hide" (otherwise we could never free or delete a buffer).
-  if (!buf->terminal) {
+  if (!buf->terminal && !buf->b_messages) {
```

Comment:
I recommend not trying to solve that problem before it actually is proven to be a problem. Even for terminal buffers, I think we're doing it wrong.

If something weird happens because the user explicitly made something  weird happen: who cares? :)